jina/peapods/peas/base.py

```diff
@@ -0,0 +1,150 @@
+import argparse
+import multiprocessing
+import os
+import threading
+
+from .helper import _get_event, _make_or_event
+from ..runtimes import Runtime
+from ... import __stop_msg__
+from ...excepts import RuntimeFailToStart, RuntimeTerminated
+from ...helper import typename
+from ...logging.logger import JinaLogger
+
+__all__ = ['BasePea']
+
+
+class PeaType(type):
+    _dct = {}
+
+    def __new__(cls, name, bases, dct):
+        _cls = super().__new__(cls, name, bases, dct)
+        PeaType._dct.update({name: {'cls': cls,
+                                    'name': name,
+                                    'bases': bases,
+                                    'dct': dct}})
+        return _cls
+
+    def __call__(cls, *args, **kwargs) -> 'PeaType':
+        # switch to the new backend
+        _cls = {
+            'thread': threading.Thread,
+            'process': multiprocessing.Process,
+        }.get(getattr(args[0], 'runtime', 'thread'))
+
+        # rebuild the class according to mro
+        for c in cls.mro()[-2::-1]:
+            arg_cls = PeaType._dct[c.__name__]['cls']
+            arg_name = PeaType._dct[c.__name__]['name']
+            arg_dct = PeaType._dct[c.__name__]['dct']
+            _cls = super().__new__(arg_cls, arg_name, (_cls,), arg_dct)
+
+        return type.__call__(_cls, *args, **kwargs)
+
+
+class BasePea(metaclass=PeaType):
+    def __init__(self, args: 'argparse.Namespace'):
+        self.args = args
+        self.name = self.args.name or self.__class__.__name__
+        self.is_ready = _get_event(self)
+        self.is_shutdown = _get_event(self)
+        self.ready_or_shutdown = _make_or_event(self, self.is_ready, self.is_shutdown)
+        self.logger = JinaLogger(self.name,
+                                 log_id=self.args.log_id,
+                                 log_config=self.args.log_config)
+        self.runtime = Runtime(args)
+
+    def run(self):
+        """ Method representing the :class:`BaseRuntime` activity.
+        """
+
+        def _finally():
+            self.is_ready.clear()
+            self.is_shutdown.set()
+            self._unset_envs()
+
+        self._set_envs()
+
+        try:
+            self.runtime.setup()
```

Comment:
if possible, using Runtime as context manager will be cleaner and more robust than a complicated set of Exceptions. 

I also think that the different Runtimes are so diferent that I think is better to have different run methods for each. It is easier to read specialized methods than fitting 3 things in the same template.

Reply:
For instance, RuntimeTerminated is only thrown by a local Runtime. Each of them will have their own nuances and I think is better to have it prepared to be of different nature, so that they can evolve in different directions.

Reply:
> Each of them will have their own nuances

The `BaseRuntime` will capture this pattern for all.

Note throwing exception `RuntimeTerminated ` is not defined as *the way* to end from `run_forever`, see `BaseRuntime` docstring: calling `cancel` will can end a forever loop peacefully. The Runtime designer must implement it.